content/en/blog/2021/simple-vms-part-2/index.md

```diff
@@ -0,0 +1,290 @@
+---
+title: "Part 2 - Implementing an easier way to add virtual machines to Istio service mesh"
+description: How to statically configure envoy to join a mesh.
+publishdate: 2022-01-17
+attribution: "Nick Nellis (Solo.io)"
+keywords: [virtualmachine,Istio,networking,envoy]
+---
+![Virtual Machine](./istio-setup.png)
+
+Part 1 - [An easier way to add virtual machines to Istio service mesh](/blog/2021/simple-vms/)
+
+A lot of Istio users would like to connect their virtual machines to their service mesh but the current process is very cumbersome and hard to automate. This series explains a simpler way to connect outside applications securely to your applications running in a service mesh. If you would like to know more about why this might work for you, check out part 1 here.
```

Comment:
It seems a bit concerning that we are calling Istio's VM operation cumbersome and recommending manually configuring Envoy, requiring deep knowledge of Istio internals that are subject to change in the future. It's not clear to me how this is simpler in any way. I don't feel great about recommending people to not verify trust domains either, at least not without sufficient warning.

Reply:
They way VMs need to be provisioned today with Istio is very difficult for a number of reasons. All of our customers who are interested in adding their VMs to Istio service meshes are asking for an alternative to the current process. We instead do have customers running the above solution in production already.

Reply:
What is so hard about the provisioning process that this is easier? I am not sure I understand this - seems like perhaps there is a massive gap. One of the main values of Istio is doing a bunch of things for you so you don't have to - like provisioning certs, configuring envoy, etc. Using alternative cert provisioning mechanisms is perfectly reasonable, but I don't see why users are wanting to manually configure envoy config directly - even the Envoy team states that XDS is *not* intended for humans.

I don't know what our guidelines are for blogs, but I do not *personally* feel good about publishing information that I think is harmful to users and gives Istio a bad reputation. I would much prefer we work towards making the current VM state easier to use.

Reply:
Here are some of the top frustrations with the current process
* Requirement from CI to communicate with the k8s API before a VM is started
* Issues on-boarding VMS in AWS
  * https://github.com/istio/istio/issues/33268
  * https://github.com/istio/istio/issues/31624
* Iptables manipulation
* No docker support
* Windows support

Reply:
All really great points - thanks! Just want to clarify as well that I was not trying to say Istio VM story is perfect right now, just that our (long term) strategy should be to improve it rather than sidestep it. Totally understand if users want to do that for now though.

> Requirement from CI to communicate with the k8s API before a VM is started

Totally agree. I think this is actually largely a documentation gap - using an external cert provisioning tool, as discussed here, is perfectly valid 

> AWS

I think this all comes down to the LoadBalancer service having hostname instead of IP? If so, https://github.com/istio/istio/pull/36422/ might resolve some or all of this. If not, worth looking into more.

> iptables

Istio does support running without iptables and using HTTP_PROXY or other mechanisms. However, its under documented (and tested!).

> Docker support

Again, I think this is largely a documentation issue. I don't think there are any technical blockers (after all, we run in docker on k8s).

> Windows

How does it work in this blog. I saw we use `image: gcr.io/istio-release/proxyv2:1.12.0`. In windows case are you using a custom istio/proxy built for windows or just using upstream Envoy?


Overall I think we can improve a lot of the VM UX through improvements to istio.io. I don't think many changes are really needed in Istio code, just to expose them to users.

Reply:
Thanks for the great responses to the issues. I agree that over time it would be great to improve Istio so this solution isn't needed but a this time a lot of people are simply choosing not to do any mesh VMs. The solution i wrote about actually came from a customer looking to lower the barrier of entry. They aren't the only customer choosing this solution over the existing Istio one. In my opinion, its more beneficial to share how people are actually using Istio than some ideal "right way". The intent of this blog is to not only display how Istio internals work, but add more creative thought on how it can utilized to provide benefit.